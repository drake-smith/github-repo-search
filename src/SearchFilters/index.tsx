import React from 'react';
import styled from 'styled-components';
import { LANGUAGES, SORT_BY } from '../shared/constants';

interface SearchFiltersProps {
  sortBy: string;
  handleSortByChange: (sortBy: string) => void;
  language: string;
  handleLanguageChange: (language: string) => void;
  isDisabled: boolean;
}

const SearchFiltersContainer = styled.div`
  margin-top: 1rem;
`;

const SortByLabel = styled.p`
  font-weight: bold;
  text-transform: uppercase;
  font-style: italic;
`;

const InputsContainer = styled.div`
  background: #f7f7f9;
  padding: 1rem;
  display: flex;
  border-radius: 10px;
`;

const RadioLabel = styled.label`
  margin-left: 0.5rem;
`;

const LanguageContainer = styled.div`
  padding-left: 2rem;
`;

const LanguageSelect = styled.select`
  display: block;
`;

const SearchFilters = ({
  sortBy,
  handleSortByChange,
  language,
  handleLanguageChange,
  isDisabled,
}: SearchFiltersProps) => {
  const languageSelects: Array<JSX.Element> = [];
  languageSelects.push(
    <option key={'none'} value={''} data-testid={'language-option'}>
      None
    </option>
  );
  for (const key in LANGUAGES) {
    languageSelects.push(
      <option key={key} value={key} data-testid={'language-option'}>
        {LANGUAGES[key]}
      </option>
    );
  }
  return (
    <SearchFiltersContainer>
      <SortByLabel>Sort by:</SortByLabel>
      <InputsContainer>
        <div>
          <div>
            <input
              type="radio"
              id="default"
              data-testid="default"
              name="Best match"
              value="best match"
              checked={sortBy === SORT_BY.default}
              onChange={() => handleSortByChange(SORT_BY.default)}
              disabled={isDisabled}
            />
            <RadioLabel htmlFor="default">Best match</RadioLabel>
          </div>
          <div>
            <input
              type="radio"
              id="stars"
              data-testid="stars"
              name="Total Stars"
              value="stars"
              checked={sortBy === SORT_BY.stars}
              onChange={() => handleSortByChange(SORT_BY.stars)}
              disabled={isDisabled}
            />
            <RadioLabel htmlFor="stars">Total stars</RadioLabel>
          </div>
        </div>
        <LanguageContainer>
          <label htmlFor="language-select">Language</label>
          <LanguageSelect
            data-testid="language-select"
            value={language}
            id="language-select"
            onChange={(e) => handleLanguageChange(e.target.value)}
            disabled={isDisabled}
          >
            {languageSelects}
          </LanguageSelect>
        </LanguageContainer>
      </InputsContainer>
    </SearchFiltersContainer>
  );
};

export default SearchFilters;
