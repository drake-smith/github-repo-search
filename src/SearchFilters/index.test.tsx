import React from 'react';
import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import SearchFilters from '.';

describe('<SearchFilters />', () => {
  const handleLanguageChange = jest.fn();
  const handleSortByChange = jest.fn();

  test('It renders correctly', () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <SearchFilters
            language={''}
            handleLanguageChange={handleLanguageChange}
            sortBy={''}
            handleSortByChange={handleSortByChange}
            isDisabled={false}
          />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('handler functions are called correctly', () => {
    render(
      <MemoryRouter>
        <SearchFilters
          language={''}
          handleLanguageChange={handleLanguageChange}
          sortBy={''}
          handleSortByChange={handleSortByChange}
          isDisabled={false}
        />
      </MemoryRouter>
    );

    const dropdown = screen.getByTestId('language-select') as HTMLSelectElement;
    expect(dropdown.value).toBe('');
    fireEvent.change(dropdown, { target: { value: 'javascript' } });
    expect(handleLanguageChange).toHaveBeenCalledTimes(1);

    const defaultRadio = screen.getByTestId('default') as HTMLInputElement;
    expect(defaultRadio.checked).toBeTruthy();

    const clickedOnRadio = screen.getByTestId('stars') as HTMLInputElement;
    fireEvent.click(clickedOnRadio);

    expect(handleSortByChange).toHaveBeenCalledTimes(1);
  });
});
