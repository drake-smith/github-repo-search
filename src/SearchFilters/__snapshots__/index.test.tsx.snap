// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchFilters /> It renders correctly 1`] = `
<div
  className="sc-bdvvtL NqmGr"
>
  <p
    className="sc-gsDKAQ RTiBz"
  >
    Sort by:
  </p>
  <div
    className="sc-dkPtRN dddJjX"
  >
    <div>
      <div>
        <input
          checked={true}
          data-testid="default"
          id="default"
          name="Best match"
          onChange={[Function]}
          type="radio"
          value="best match"
        />
        <label
          className="sc-hKwDye ggeeAH"
          htmlFor="default"
        >
          Best match
        </label>
      </div>
      <div>
        <input
          checked={false}
          data-testid="stars"
          id="stars"
          name="Total Stars"
          onChange={[Function]}
          type="radio"
          value="stars"
        />
        <label
          className="sc-hKwDye ggeeAH"
          htmlFor="stars"
        >
          Total stars
        </label>
      </div>
    </div>
    <div
      className="sc-eCImPb ellptx"
    >
      <label
        htmlFor="language-select"
      >
        Language
      </label>
      <select
        className="sc-jRQBWg iUJxXW"
        data-testid="language-select"
        id="language-select"
        onChange={[Function]}
        value=""
      >
        <option
          data-testid="language-option"
          value=""
        >
          None
        </option>
        <option
          data-testid="language-option"
          value="javascript"
        >
          JavaScript
        </option>
        <option
          data-testid="language-option"
          value="typescript"
        >
          TypeScript
        </option>
        <option
          data-testid="language-option"
          value="python"
        >
          Python
        </option>
        <option
          data-testid="language-option"
          value="ruby"
        >
          Ruby
        </option>
        <option
          data-testid="language-option"
          value="java"
        >
          Java
        </option>
        <option
          data-testid="language-option"
          value="php"
        >
          PHP
        </option>
        <option
          data-testid="language-option"
          value="c#"
        >
          C#
        </option>
        <option
          data-testid="language-option"
          value="c++"
        >
          C++
        </option>
      </select>
    </div>
  </div>
</div>
`;
